### PR TITLE
[node server] render line, bar, map, and scatter tiles

### PR DESCRIPTION
**Pull request by chejennifer**
screen-recording-_95_

### CODE REVIEW

File: static/src/server.ts [hunk 1]
```diff
@@ -15,13 +15,76 @@
...
+} from "../js/chart/draw_map_utils";
+import { SELF_PLACE_DCID_PLACEHOLDER } from "../js/constants/subject_page_constants";
+import { NamedTypedPlace, StatVarSpec } from "../js/shared/types";
+import { TileConfig } from "../js/types/subject_page_proto_types";
 
 const app = express();
 const port = process.env.PORT || 3030;
+// Each value in the array is the width of the character with ascii code of
+// array index + 32 for 10px Roboto font.
+const CHAR_WIDTHS = [
```
Comment (shifucun) [inline static/src/server.ts:55]: Add comments on how to get this array?

File: static/src/server.ts [hunk 2]
```diff
@@ -15,13 +15,76 @@
...
+  5.6875, 5.53125, 6.8125, 7.1328125, 2.7265625, 5.5234375, 6.2734375,
+  5.3828125, 8.734375, 7.1328125, 6.875, 6.3125, 6.875, 6.1640625, 5.9375,
+  5.96875, 6.484375, 6.3671875, 8.875, 6.2734375, 6.0078125, 5.9921875, 2.65625,
+  4.1015625, 2.65625, 4.1796875, 4.515625, 3.09375, 5.4453125, 5.6171875,
+  5.234375, 5.640625, 5.3046875, 3.359375, 5.6171875, 5.5078125, 2.4296875,
+  2.390625, 5.0703125, 2.4296875, 8.765625, 5.5234375, 5.703125, 5.6171875,
+  5.6875, 3.390625, 5.15625, 3.2734375, 5.515625, 4.84375, 7.515625, 4.9609375,
+  4.734375, 4.9609375, 3.390625, 2.4375, 3.390625, 6.8046875, 0,
+];
+// Average width of a 10px Roboto character.
```
Comment (shifucun) [inline static/src/server.ts:70]: how to get this value?

**Review by shifucun — APPROVED** [review]
this is very very cool!

**Review by pradh — COMMENTED** [review]
This is super cool Jennifer!!

File: static/js/chart/draw_map_utils.ts [hunk 3]
```diff
@@ -29,12 +29,13 @@ const MIN_COLOR = "#f0f0f0";
...
-const LEGEND_MARGIN_RIGHT = 5;
-const LEGEND_IMG_WIDTH = 10;
+export const LEGEND_MARGIN_VERTICAL = 6;
+export const LEGEND_MARGIN_RIGHT = 5;
+export const LEGEND_IMG_WIDTH = 10;
 const NUM_TICKS = 4;
 export const HOVER_HIGHLIGHTED_CLASS_NAME = "region-highlighted";
 const LEGEND_CLASS_NAME = "legend";
-const LEGEND_TICK_LABEL_MARGIN = 10;
+export const LEGEND_TICK_LABEL_MARGIN = 10;
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/chart/draw_map_utils.ts]: done

File: static/js/components/tiles/map_tile.tsx [hunk 4]
```diff
@@ -167,41 +171,41 @@ export function MapTile(props: MapTilePropType): JSX.Element {
...
 export const fetchData = async (
-  place: NamedTypedPlace,
-  enclosedPlaceType: string,
-  statVarSpec: StatVarSpec
+  props: MapTilePropType
 ): Promise<MapChartData> => {
   const geoJsonPromise = axios
     .get(
-      `/api/choropleth/geojson?placeDcid=${place.dcid}&placeType=${enclosedPlaceType}`
+      `${props.apiRoot || ""}/api/choropleth/geojson?placeDcid=${
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/components/tiles/map_tile.tsx]: done

File: static/js/components/tiles/scatter_tile.tsx [hunk 5]
```diff
@@ -201,27 +200,31 @@ function getPopulationPromise(
...
-    statVarSpec
+    props.place.dcid,
+    props.enclosedPlaceType,
+    props.statVarSpec,
+    props.apiRoot
   );
   const placeNamesPromise = axios
     .get(
-      `/api/place/descendent/name?dcid=${placeDcid}&descendentType=${enclosedPlaceType}`
+      `${props.apiRoot || ""}/api/place/descendent/name?dcid=${
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/components/tiles/scatter_tile.tsx]: done

File: static/src/server.ts (hunk 1 above)
Comment (chejennifer) [reply static/src/server.ts:60]: done

File: static/src/server.ts (hunk 2 above)
Comment (chejennifer) [reply static/src/server.ts:75]: done

**Review by shifucun — APPROVED** [review]

**Pull request merged**